File: clover.nim
import db_sqlite
import future
import os
import parsecsv
import re
import streams
import strutils
import tables
import unicode

proc opendb() : DbConn =
  open("clover.db", nil, nil, nil)

type Datatype = enum
  Integer, Boolean, Text
  
type Format = tuple[column_name: string, width: int, datatype: Datatype]

# TODO: Not a real titlecase function, but works for us.
proc toTitleAscii(s: string) : string =
  s[0].toUpperAscii() & s[1..<s.len].toLowerAscii()

proc parseDatatype(name: string) : Datatype =
  parseEnum[Datatype](name.toTitleAscii())

proc parseRow(c: var CsvParser) : Format =
  let name = c.rowEntry("column name")
  let width = parseInt(c.rowEntry("width"))
  let datatype = parseDatatype(c.rowEntry("datatype"))
  (column_name: name, width: width, datatype: datatype)

proc readFormat(format: string) : seq[Format] =
  let path = "specs/" & format & ".csv"
  var c: CsvParser
  c.open(path)
  c.readHeaderRow()
  # TODO: Use header row instead of assuming order
  # TODO: Is there an idiomatic way to turn a while c.readRow() loop
  # into something that can be iterated in a comprehension (or otherwise
  # used immutably)? Or, alternatively, should readFormat be an iterator?
  var rows: seq[Format] = @[]
  while c.readRow():
    rows.add(parseRow(c))
  rows

proc makeTable(db: DbConn, tablename: string, format: seq[Format]) =
  # TODO: Comprehension
  var cols: seq[string] = @[]
  for col in format:
    let colstr = col.column_name & " " & $col.datatype
    cols.add(colstr)
  let colsbit = join(cols, ",\n")
  let query = "CREATE TABLE IF NOT EXISTS " & tablename & "(" & colsbit & ")"
  db.exec(sql(query))

# TODO: Iterate sequences (or, better, tables indexed by column name?) of
# (string or int or bool) instead of converting then converting back to string
iterator iterDataFile(fname: string, format: seq[Format]) : seq[string] =
  var s = newFileStream(fname, fmRead)
  if not isNil(s):
    var line = ""
    while s.readline(line):
      var pos = 0
      var row: seq[string] = @[]
      let uline = toRunes(line)
      for col in format:
        # TODO: There has to be a better way to slice unicode?
        let uvalue = uline[pos..<min(uline.len, pos+col.width)]
        let svalue = $uvalue;
        let value = svalue.strip(leading=false)
        case col.datatype
        of Text: row.add(value)
        of Integer: row.add($(parseInt(value.strip())))
        of Boolean: row.add($(parseInt(value.strip()))) # 0 or 1, not False or True
        pos = pos + col.width
      yield row
    s.close()

let rfname = re"data/(\w+?)_\d\d\d\d-\d\d-\d\d\.txt"
proc handleDataFile(db: DbConn, fname: string) =
  var matches: array[1, string]
  if fname.match(rfname, matches):
    let formatName = matches[0]
    let format = readFormat(formatName)
    makeTable(db, formatName, format)
    db.exec(sql"BEGIN")
    try:
      # TODO: comprehension
      var cols: seq[string] = @[]
      var vals: seq[string] = @[]
      for col in format:
        cols.add(col.column_name)
        vals.add("?")
      let query = "INSERT INTO " & formatName & " (" & join(cols, ", ") &
        ") VALUES (" & join(vals, ", ") & ")"
      echo query
      for row in iterDataFile(fname, format):
        echo row
        db.exec(sql(query), row)
      db.exec(sql"COMMIT")
    except:
      echo $(getCurrentException().name) & ": " & getCurrentExceptionMsg()
      db.exec(sql"ROLLBACK")
    
let db = opendb()
for fname in walkFiles("data/*.txt"):
  echo fname
  handleDataFile(db, fname)
for row in db.fastRows(sql"SELECT * FROM superheroes"):
  echo row
db.close()

File: clover.py
#!/usr/bin/env python3

import csv
import logging
import os
import re
import sqlite3
import sys

def opendb():
    return sqlite3.connect('clover.db')

def read_format(format):
    with open(os.path.join('specs', f'{format}.csv'), encoding='utf-8') as f:
        for col in csv.DictReader(f):
            col['width'] = int(col['width'])
            yield col

_CONVERSIONS = {
    'INTEGER': int,
    'BOOLEAN': lambda x: bool(int(x)),
    'TEXT': str
    }
            
def read_datafile(fname, format):
    with open(fname, encoding='utf-8') as f:
        for line in f:
            logger.debug(f"'{line.rstrip()}'")
            pos = 0
            row = {}
            for col in format:
                value = line[pos:pos+col['width']].rstrip()
                value = _CONVERSIONS[col['datatype']](value)
                row[col['column name']] = value
                pos += col['width']
            yield row

def make_table(db, tablename, format):
    cols = ',\n'.join('{column name} {datatype}'.format_map(col) for col in format)
    sql = f'''CREATE TABLE IF NOT EXISTS {tablename} ({cols})'''
    logging.debug(f"'{sql}'")
    db.execute(sql)

rfname = re.compile(r'data/([A-Za-z0-9]+?)_(\d\d\d\d-\d\d-\d\d)\.txt')
def handle_datafile(db, fname):
    logger.info(f"data file '{fname}'")
    try:
        m = rfname.match(fname)
        formatname = m.group(1)
        logger.debug(f"reading format '{formatname}'")
        format = list(read_format(formatname))
        make_table(db, formatname, format)
        logger.debug("'BEGIN'")
        db.execute('BEGIN')
        cols = ', '.join(col['column name'] for col in format)
        vars = ', '.join('?' for col in format)
        sql = f'INSERT INTO {formatname} ({cols}) VALUES ({vars})'
        # NOTE: It might be slightly more efficient to use executemany here,
        # but this way we can log each row.
        try:
            for row in read_datafile(fname, format):
                values = list(row.values())
                logger.debug(f"'{sql}', {values}")
                db.execute(sql, values)
        except:
            logger.debug("'ROLLBACK'")
            db.execute('ROLLBACK')
        else:
            logger.debug("'COMMIT'")
            db.execute('COMMIT')
    except Exception as e:
        logger.exception(f"failed on data file '{fname}'")

def handle_dir(db, dname):
    for fname in os.listdir(dname):
        if fname.endswith('.txt'):
            handle_datafile(db, os.path.join(dname, fname))
 
if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG,
                        filename='clover.log',
                        format='%(asctime)s:%(levelname)-8s:%(message)s')
    logger = logging.getLogger('clover')
    console = logging.StreamHandler()
    console.setLevel(logging.INFO)
    logger.addHandler(console)
    path = os.path.abspath(os.path.dirname(sys.argv[0]))
    os.chdir(path)
    with opendb() as db:
        handle_dir(db, 'data')
